app/src/main/java/unipd/se18/ocrcamera/MignifyResolver.java

```diff
@@ -0,0 +1,111 @@
+                    //Accept and Content-Type selection header
+                    .header("Accept", "application/json")
+                    //Type of response we expect
+                    .asJson();
+        } catch (UnirestException e) {
+            Log.e("Unirest error", e.getMessage());
+        }
+        if (response == null) {
+            product = HTTP_ERROR;
+        } else {
+            try {
+                //Retrieve the parsed JSONObject from the Mashape response
+                JSONObject jsonResponse = response.getBody().getObject();
+                //Getting the status from the JSON
+                //The server responds with either "ok" or "error"
+                String status = jsonResponse.getString("status");
+                if (status.equals("ok")) {
+                    //If the EAN code was found, in the JSON there is an list of products, each
+                    //one with name, brand and language
+                    JSONArray productList = jsonResponse.getJSONArray("results");
+                    String[] productNames = new String[productList.length()];
+                    //Future proofing usage of brands and language, still unimplemented
+                    String[] productBrands = new String[productList.length()];
+                    String[] productLanguages = new String[productList.length()];
+                    //If no product are in the response
+                    if (jsonResponse.length() == 0) {
+                        product = NO_PRODUCT_ERROR;
+                    } else {
+                        //Getting name, brand and language for each product
+                        for (int i = 0; i < jsonResponse.length(); i++) {
+                            //Getting the current product from the JSON list and it's values
+                            JSONObject currentJsonObject = productList.getJSONObject(i);
+                            productNames[i] = currentJsonObject.getString("productName");
+                            productBrands[i] = currentJsonObject.getString("brand");
+                            productLanguages[i] = currentJsonObject.getString("languageCode");
+                        }
+
+                        //Selecting the first name in the list as the result product name.
+                        //It should be the most plausible product name.
+                        //In the future a better selection will be made, checking brand and language
+                        product = productNames[0];
+                    }
+                } else if(status.equals("error")){
+                    //In case the EAN wasn't valid, the error in the JSON will be returned
+                    product = jsonResponse.getString("message");
+                }
+            } catch (JSONException e){
+                product = JSON_ERROR;
+                Log.e("JSON error", e.getMessage());
+            }
+        }
+        return product;
+    }
+}
```

Comment:
In general in this method there's a lot of hardcoding. If I were you, I've stored all JSON fields' keys into some constants or, even better, I've used android string resources to save them. In this way if in the future these names will change, you have to modify them in one place and you're not forced to look through the whole code.

Reply:
Pietro Prandini was kind enough to suggest to use string.xml for this usage.
So I do agree with the both of you, I should move the JSON and fields and error constants to the string file.

Reply:
Since this is a Module, and it doesn't have the context, I cannot access the local string.xml, but only the global one with `Resources.getSystem().getString(android.R.string.supercool)`, correct me if I'm wrong.
So for the moment I'll put everything meaningful in constants.